apps/condo/domains/miniapp/schema/B2BAccessToken.js

```diff
@@ -0,0 +1,282 @@
+
+function getValidationError ({
+    b2bAppContext,
+    accessRightSet,
+    accessRightForUserAndApp,
+}) {
+    if (!b2bAppContext) {
+        return ERRORS.CONTEXT_NOT_EXISTS
+    }
+    if (!accessRightSet) {
+        return ERRORS.RIGHT_SET_NOT_EXISTS
+    }
+    if (get(accessRightSet, 'type') !== ACCESS_RIGHT_SET_SCOPED_TYPE) {
+        return ERRORS.WRONG_RIGHT_SET_TYPE
+    }
+    if (get(accessRightSet, 'app') && (get(accessRightSet, 'app') !== get(b2bAppContext, 'app'))) {
+        return ERRORS.CONTEXT_DOES_NOT_MATCH_RIGHT_SET
+    }
+    if (!accessRightForUserAndApp) {
+        return ERRORS.SERVICE_USER_DOES_NOT_MATCH_CONTEXT
+    }
+    return null
+}
+
+const ERRORS = {
+    CONTEXT_NOT_EXISTS: {
+        code: BAD_USER_INPUT,
+        type: ACCESS_TOKEN_CONTEXT_NOT_EXISTS,
+        message: 'B2BAppContext does not exists',
+    },
+    RIGHT_SET_NOT_EXISTS: {
+        code: BAD_USER_INPUT,
+        type: ACCESS_TOKEN_RIGHT_SET_NOT_EXISTS,
+        message: 'B2BAppAccessRightSet does not exists',
+    },
+    WRONG_RIGHT_SET_TYPE: {
+        code: BAD_USER_INPUT,
+        type: ACCESS_TOKEN_WRONG_RIGHT_SET_TYPE,
+        message: 'Wrong B2BAppAccessRightSet type, must be "SCOPED"',
+    },
+    CONTEXT_DOES_NOT_MATCH_RIGHT_SET: {
+        code: BAD_USER_INPUT,
+        type: ACCESS_TOKEN_CONTEXT_DOES_NOT_MATCH_RIGHT_SET,
+        message: 'B2BAppContext and B2BRightSet connected to different B2BApps',
+    },
+    SERVICE_USER_DOES_NOT_MATCH_CONTEXT: {
+        code: BAD_USER_INPUT,
+        type: ACCESS_TOKEN_SERVICE_USER_DOES_NOT_MATCH_CONTEXT,
+        message: 'User and B2BAppContext connected to different B2BApps',
+    },
+    NEGATIVE_TTL: {
+        code: BAD_USER_INPUT,
+        type: ACCESS_TOKEN_NEGATIVE_TTL,
+        message: 'TTL can\'t be negative. You provided expiresAt datetime previous to the current one',
+    },
+}
+
+const ERRORS_FOR_DELETION = new Set([
+    ERRORS.CONTEXT_NOT_EXISTS,
+    ERRORS.RIGHT_SET_NOT_EXISTS,
+    ERRORS.NEGATIVE_TTL,
+])
+
+const B2BAccessToken = new GQLListSchema('B2BAccessToken', {
+    schemaDoc: 'Access to service user with additional restrictions by Organization and B2BAppAccessRightSet',
+    fields: {
+        sessionId: {
+            schemaDoc: 'Encrypted sessionId of session',
+            type: 'EncryptedText',
+            isRequired: true,
+            access: access.updatableOnlyForAdmin,
+            hooks: {
+                resolveInput ({ operation, existingItem, fieldPath, originalInput }) {
+                    const inputSessionId = get(originalInput, fieldPath)
+                    // need to have option to manually set sessionId, so we can migrate already given tokens to this
+                    if (!isNil(inputSessionId) && !isEmpty(inputSessionId)) {
+                        return inputSessionId
+                    }
+                    if (operation === 'create') {
+                        return ACCESS_TOKEN_SESSION_ID_PREFIX + crypto.randomUUID()
```

Comment:
default behaviour?

Reply:
I found out that in case of virtual field, only 'read' access is called. And it does not contain information about operation (creating entity, 'create' access is not called, only 'read')